packages/devtools/lib/src/performance/performance_screen.dart

```diff
@@ -0,0 +1,234 @@
+      ..click(_clear);
+
+    profilingInstructions = div(c: 'message')
+      ..layoutVertical()
+      ..flex()
+      ..add([
+        div(c: 'instruction-message')
+          ..layoutHorizontal()
+          ..flex()
+          ..add([
+            div(text: 'Click the record button '),
+            createIconElement(instructionsRecord),
+            div(text: 'to start recording a CPU profile.')
+          ]),
+        div(c: 'instruction-message')
+          ..layoutHorizontal()
+          ..flex()
+          ..add([
+            div(text: 'Click the stop button '),
+            createIconElement(stop),
+            div(text: 'to end the recording.')
+          ]),
+      ]);
+
+    recordingMessage = div(c: 'message')
+      ..layoutVertical()
+      ..flex()
+      ..add([
+        div(text: 'Recording', c: 'recording-message'),
+        Spinner.centered(),
+      ]);
+
+    cpuProfiler = _CpuProfiler(performanceController);
+
+    tabNav = CpuProfilerTabNav(
+      cpuProfiler,
+      CpuProfilerTabOrder(
+        first: CpuProfilerViewType.callTree,
+        second: CpuProfilerViewType.bottomUp,
+        third: CpuProfilerViewType.flameChart,
+      ),
+    );
+  }
+
+  void _startRecording() {
+    performanceController.startRecording();
+    _updateCpuProfilerVisibility(hidden: true);
+    _updateButtonStates();
+    profilingInstructions.hidden(true);
+    recordingMessage.hidden(false);
+  }
+
+  void _stopRecording() async {
+    performanceController.stopRecording();
+    recordingMessage.hidden(true);
+    _updateCpuProfilerVisibility(hidden: false);
+    _updateButtonStates();
+    await cpuProfiler.update();
+  }
+
+  void _clear() {
+    performanceController.reset();
+    _updateCpuProfilerVisibility(hidden: true);
+    profilingInstructions.hidden(false);
+  }
+
+  void _updateButtonStates() {
+    startRecordingButton.disabled = performanceController.recording;
+    clearButton.disabled = performanceController.recording;
+    stopRecordingButton.disabled = !performanceController.recording;
+  }
+
+  void _updateCpuProfilerVisibility({@required bool hidden}) {
+    tabNav.element.hidden(hidden);
+    cpuProfiler.hidden(hidden);
+  }
+}
+
+class _CpuProfiler extends CpuProfiler {
+  _CpuProfiler(this._performanceController)
+      : super(
+          _CpuFlameChart(_performanceController),
+          _CpuCallTree(_performanceController),
+          _CpuBottomUp(_performanceController),
+          defaultView: CpuProfilerViewType.callTree,
+        );
+
+  final PerformanceController _performanceController;
+
+  @override
+  Future<void> prepareCpuProfile() async {
+    _performanceController.mergeRecordedProfiles();
+  }
+
+  @override
+  bool maybeShowMessageOnUpdate() {
+    if (_performanceController.cpuProfileData == null) {
+      showMessage(div(text: 'No CPU samples recorded.'));
+      return true;
+    }
+    return false;
+  }
+}
+
+class _CpuFlameChart extends CpuFlameChart {
```

Comment:
avoid creating these classes by refactoring so CpuFlameChart can be used for both. Same for the other cases. For example, having a shared model base class could solve this or another way of abstracting without subclasses would be cleaner.

Reply:
Used a data provider. Fixed.